### PR TITLE
Booking more progress on SI-6666

### PR DESCRIPTION
**Pull request by retronym**
- Avoids unneeded outer pointers for classes nested indirectly in self/super calls
- Expands implementation restriction to prevent verify errors in early initializers

Closes SI-6259 and SI-6506; a bit more work is needed on SI-6666/SI-6957.

Review by @paulp / @JamesIry.

I'll warn you up front that there is some less than beautiful code in `ExplicitOuter` to communicate these errors as implementation restrictions, rather than crashes.

### CODE REVIEW

**Comment by retronym** [comment]
```
ticket/6666 ~/code/scala2 git log --oneline head~1...
1ee0756 Class symbols can't be contravariant.

ticket/6666 ~/code/scala2 ./tools/partest-ack t4842
Found 2 tests matching 'ack t4842'

Testing individual files
testing: [...]/files/pos/t4842.scala                                  [  OK  ]

Testing individual files
testing: [...]/files/neg/t4842.scala                                  [  OK  ]
All of 2 tests were successful (elapsed time: 00:00:03)

ticket/6666 ~/code/scala2 SCALAC_OPTS=-optimize ./tools/partest-ack t4842
Found 2 tests matching 'ack t4842'

Testing individual files
testing: [...]/files/pos/t4842.scala                                  [  OK  ]

Testing individual files
testing: [...]/files/neg/t4842.scala                                  [  OK  ]
All of 2 tests were successful (elapsed time: 00:00:05)
```

@lrytz  It would be useful if the build script showed the hash of the PR branch that it builds.

Right now we see:

```
From github.com:scala/scala
 * [new ref]         refs/pull/2017/head -> pullreq/2017
+ git merge pullreq/2017
```

Desired:

```
From github.com:scala/scala
 * [new ref]         refs/pull/2017/head -> pullreq/2017
+ git describe pullreq/2017
v2.10.0-319-g1ee0756
+ git merge pullreq/2017
```

**Comment by lrytz** [comment]
The information is in the output in a few places, e.g.

```
+ revision=ffee165870a442e20dcaef4898732f255a0a3083
```

or

```
     [echo]          git hash: ffee165870
```

should i add it anyway?

**Comment by retronym** [comment]
I think that's the hash of the merge commit; I'm looking for confirmation that it is building 1ee0756, the latest commit in the PR.

**Comment by lrytz** [comment]
oh right..! i updated the scripts.

**Comment by adriaanm** [comment]
files/run/class-symbol-contravariant.scala [FAILED]

**Comment by retronym** [comment]
Okay, kitteh, the forgotten checkfile is now available. (censored), and fly!
happy_kitten_by_idiotmee-d5c82x7

**Comment by retronym** [comment]
@adriaanm @lrytz methinks the jenkins box is overloaded

**Comment by lrytz** [comment]
yes indeed. we have to collect a bit of data, but it seems that i have too many slaves running on "chara" at least

**Comment by lrytz** [comment]
we're running 4 concurrent jobs, the machine has 8 cores and 8 gigs of memory, so in principle it should be ok.. but the processor is a bit old (introduced 2006, intel X5355)

**Comment by paulp** [comment]
"we're running 4 concurrent jobs, the machine has 8 cores and 8 gigs of memory, so in principle it should be ok.."

In my experience that is heavily overtaxed. That is up to 32 simultaneous tests (partest will be running 8 on each concurrent job.) My 8-core machine has 16 GB and allowing jenkins 4 executors sends the load over 20.

**Comment by paulp** [comment]
...and since that alone is not surprising, let me clarify it's the kind of 20 you can really feel, as in the machine has been ground to a halt.

**Comment by lrytz** [comment]
partest is not running in all builds at the same time, we spend more time compiling and scaladoc'ing

in any case, i reduced the number of slaves on that machine to 2.

**Comment by adriaanm** [comment]
BUILDLOG?

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 1]
```diff
@@ -123,9 +123,12 @@ trait Namers extends MethodSynthesis {
...
       setPrivateWithin(tree, sym, tree.mods)
 
-    def inConstructorFlag: Long =
-      if (owner.isConstructor && !context.inConstructorSuffix || owner.isEarlyInitialized) INCONSTRUCTOR
-      else 0l
+    def inConstructorFlag: Long = {
+      val termOwnedContexts: List[Context] = context.enclosingContextChain.takeWhile(_.owner.isTerm)
+      val constructorNonSuffix = termOwnedContexts exists (c => c.owner.isConstructor && !c.inConstructorSuffix)
+      val earlyInit            = termOwnedContexts exists (_.owner.isEarlyInitialized)
+      if (constructorNonSuffix || earlyInit) INCONSTRUCTOR else 0L
```
Comment (JamesIry) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:130]: Slightly prefer

def earlyInitOrConstructorNonSuffix(c: Context) => (c.owner.isConstructor && !c.inConstrucorSuffix) || c.owner.isEarlyInitialized
if (termOwnedContexts exists earlyInitOrConstructorNonSuffix) INCONSTRUCTOR else 0L

But not enough to hold up the merge.

**Comment by JamesIry** [comment]
LGTM but waiting on @paulp for review before merge

**Comment by paulp** [comment]
LGTM. I got caught up chasing down some sick stuff this revealed to me, but it's not directly related.

**Pull request merged**